File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int res = 0;
        Set<Character> set = new HashSet<>();
        int l = 0;
        for(int i = 0; i < s.length(); i++) {
            char c1 = s.charAt(i);
            while(!set.add(c1)) {
                char c2 = s.charAt(l);
                set.remove(c2);
                l++;
            }
            res = Math.max(res, i - l + 1);
        }
        return res;
    }
}

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n, left, res = len(s), 0, 0
        dic = {}
        
        for i in range(n):
            c = s[i]
            while c in dic.keys():
                del dic[s[left]]
                left += 1
            dic[c] = i 
            
            res = max(res, i - left + 1)
        return res